nomos-services/blend/src/core/mod.rs

```diff
@@ -695,77 +730,132 @@ async fn run_event_loop<
+                recovery_checkpoint = handle_local_data_message(local_data_message, &mut crypto_processor, message_scheduler, recovery_checkpoint).await;
             }
             Some(incoming_message) = blend_messages.next() => {
-                current_recovery_checkpoint = handle_incoming_blend_message(incoming_message, &mut message_scheduler, &crypto_processor, &mut blending_token_collector, current_recovery_checkpoint);
+                recovery_checkpoint = handle_incoming_blend_message(incoming_message, message_scheduler, &crypto_processor, old_session_crypto_processor.as_ref(), blending_token_collector, recovery_checkpoint);
             }
             Some(round_info) = message_scheduler.next() => {
-                current_recovery_checkpoint = handle_release_round(round_info, &mut crypto_processor, &mut rng, &backend, &network_adapter, current_recovery_checkpoint).await;
+                recovery_checkpoint = handle_release_round(round_info, &mut crypto_processor, rng, backend, network_adapter, recovery_checkpoint).await;
             }
             Some(clock_tick) = remaining_clock_stream.next() => {
-                current_public_info = handle_clock_event(clock_tick, blend_config, &mut epoch_handler, &mut crypto_processor, &mut backend, current_public_info).await;
+                public_info = handle_clock_event(clock_tick, blend_config, epoch_handler, &mut crypto_processor, backend, public_info).await;
             }
             Some(pol_info) = secret_pol_info_stream.next() => {
                 handle_new_secret_epoch_info(pol_info.poq_private_inputs, &mut crypto_processor);
             }
             Some(session_event) = remaining_session_stream.next() => {
-                match handle_session_event(session_event, blend_config, crypto_processor, current_public_info, current_recovery_checkpoint, &mut blending_token_collector, &mut backend).await {
-                    Ok(output) => {
-                        crypto_processor = output.crypto_processor;
-                        current_public_info = output.public_info;
-                        current_recovery_checkpoint = output.recovery_checkpoint;
-                    }
-                    Err((e @ (Error::LocalIsNotCoreNode | Error::NetworkIsTooSmall(_)), output)) => {
-                        tracing::info!(
-                            target: LOG_TARGET,
-                            "New membership does not satisfy the core node condition: {e:?}:  Exiting from the main event loop",
+                match handle_session_event(session_event, blend_config, crypto_processor, public_info, recovery_checkpoint, blending_token_collector, backend).await {
+                    HandleSessionEventOutput::Transitioning { new_crypto_processor, old_crypto_processor, new_public_info, new_recovery_checkpoint } => {
```

Comment:
If we are passing the `new_public_info` here, we should avoid passing components that are also created from those public info, and we should instead create them inside using the public info received, else we risk getting out of sync between those components and the public info themselves. Let's just create everything internally and have the output just return the public info, and perhaps the stuff belonging to the old session, although we already have a reference to it in here, so I don't understand why we need it to be part of the event payload?

Reply:
Also I don't think the output should have knowledge about recovery state. Can't it just tell us what the new session values are?